### PR TITLE
feat: optimize WAN 2.2 SFT

### PR DESCRIPTION
**Pull request by guozhen1997**
### Description

1. enable torch compile @thereAreDemonsNearby 

2. wrap encoder @thereAreDemonsNearby 

3. enable batch encoder @guozhen1997 

### How has this been tested?

mbs1 step time 1.7 -> 1.2 speedup 29.4%
image

mbs2 step time 4 -> 1.3 speedup 67.5%
image

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Documentation update (Document-only update)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

### Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: examples/sft/config/droid_sft_dreamzero.yaml [hunk 1]
```diff
@@ -65,16 +65,16 @@ actor:
...
-    add_value_head: false
+    add_value_head: False
     is_lora: False
     embodiment_tag: "oxe_droid"
-    relative_action: true
-    relative_action_per_horizon: false
-
+    relative_action: True
+    relative_action_per_horizon: False
+    gradient_checkpointing: False
```
Comment (qurakchin) [inline examples/sft/config/droid_sft_dreamzero.yaml:77]: why add gradient_checkpointing simultaneously in actor.model and actor.fsdp_config

File: rlinf/hybrid_engines/fsdp/utils.py [hunk 2]
```diff
@@ -421,13 +454,23 @@ def apply_fsdp2_to_model(
...
-        reshard_after_forward=False,
-    )
+    root_kwargs = {
+        "mesh": device_mesh,
+        "mp_policy": mp_policy,
+        "offload_policy": offload_policy,
+        "reshard_after_forward": False,
+    }
+
+    if _fsdp2_fully_shard_supports_ignored_params():
```
Comment (qurakchin) [inline rlinf/hybrid_engines/fsdp/utils.py:465]: if ignored_module_classes is exist but version < 2.7, this patch will not apply and no warning. please add a check, if ignored_module_classes exists but version < 2.7, raise an error to remind user.

File: rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+from einops import rearrange, repeat
+from groot.vla.model.dreamzero.modules import wan_video_vae as _groot_vae
+from tqdm import tqdm
+
+VideoVAE_ = _groot_vae.VideoVAE_
+VideoVAE38_ = _groot_vae.VideoVAE38_
+
+
+class WanVideoVAE(nn.Module):
+    def __init__(self, z_dim=16, vae_pretrained_path: str | None = None):
```
Comment (qurakchin) [inline rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py:32]: add comments to describe the optimization.

File: rlinf/models/embodiment/dreamzero/patch/wan_self_attention.py [hunk 4]
```diff
@@ -0,0 +1,328 @@
...
+
+"""Compiled helpers patched onto groot ``CausalWanSelfAttention`` (see ``get_model``)."""
+
+from __future__ import annotations
+
+import torch
+
+
+@torch.compile(mode="reduce-overhead")
+def _process_clean_image_only(
```
Comment (qurakchin) [inline rlinf/models/embodiment/dreamzero/patch/wan_self_attention.py]: if the content is same but the only diff is adding the torch.compile. use patcher.add_wrapper to do patch

File: rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+        if length <= tile:
+            return [0]
+        last_start = length - tile
+        starts = list(range(0, last_start + 1, stride))
+        if starts[-1] != last_start:
+            starts.append(last_start)
+        return starts
+
+    @staticmethod
+    def _build_spatial_tile_tasks(
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py:118]: This method seems to be a little different with the logic of the original loop in edge cases.  Is this safe?

File: rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+    clean_x=None,
+    aug_t=None,
+    y=None,
+    clip_feature=None,
+    action=None,
+    state=None,
+    embodiment_id=None,
+):
+    # Minimally-edited copy of DreamZero's CausalWanModel._forward_train.
+    # Supports micro-batch (B) > 1 with gradient checkpointing (see use_ckpt below).
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py]: maybe should say:  forbid gradient checkpointing when mbs > 1

File: rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py [hunk 7]
```diff
@@ -0,0 +1,395 @@
...
+        return x
+
+    def single_decode(self, hidden_state):
+        video = self.model.decode(hidden_state, self.scale)
+        return video.clamp_(-1, 1)
+
+    def encode(self, videos, tiled=False, tile_size=(34, 34), tile_stride=(18, 16)):
+        if videos.dim() == 4:
+            videos = videos.unsqueeze(0)
+        if tiled:
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py:250]: currently we only use tiled==false, right? (This is a question, not a modification requirement)

File: rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py [hunk 8]
```diff
@@ -0,0 +1,395 @@
...
+            video = self.tiled_decode(hidden_states, tile_size, tile_stride)
+        else:
+            video = self.single_decode(hidden_states)
+        return video
+
+    @staticmethod
+    def state_dict_converter():
+        return WanVideoVAEStateDictConverter()
+
+
```
Comment (thereAreDemonsNearby) [inline rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py:267]: Seems WanVideoVAEStateDictConverter and WanVideoVAE38 are the same as the original ones. Shouldn't patch them.

File: examples/sft/config/droid_sft_dreamzero.yaml (hunk 1 above)
Comment (thereAreDemonsNearby) [reply examples/sft/config/droid_sft_dreamzero.yaml:77]: Because our fsdp_config.gradient_checkpointing is only applicable to huggingface models. And in get_model only cfg.actor.model can be seen, while cfg.actor.fsdp_config cannot be accessed

File: rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py (hunk 8 above)
Comment (thereAreDemonsNearby) [reply rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py:267]: Now I know that money patch can  break the inheritance chain, so this is necessary for now.

File: rlinf/hybrid_engines/fsdp/utils.py (hunk 2 above)
Comment (thereAreDemonsNearby) [reply rlinf/hybrid_engines/fsdp/utils.py:465]: will add a check for this

File: rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py (hunk 6 above)
Comment (thereAreDemonsNearby) [reply rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py]: comments will be refined soon

File: rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py (hunk 5 above)
Comment (guozhen1997) [reply rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py:118]: Yes—there is indeed a subtle difference at the boundaries: the current code forces the last tile to align with the edge, ensuring complete coverage, while the original loop might miss edges. This difference is intentional and safe as long as the encoding/decoding uses the same tile strategy.

Furthermore, the bug in the original code that used a for loop to execute `tiled_encode` has been fixed in the current code. Although `tiled_encode` is not currently used.

File: rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py (hunk 6 above)
Comment (guozhen1997) [reply rlinf/models/embodiment/dreamzero/patch/wan_causal_model_forward_train.py]: thanks

File: rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py (hunk 7 above)
Comment (guozhen1997) [reply rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py:250]: yes

File: rlinf/models/embodiment/dreamzero/patch/wan_self_attention.py (hunk 4 above)
Comment (guozhen1997) [reply rlinf/models/embodiment/dreamzero/patch/wan_self_attention.py]: done

File: rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py (hunk 3 above)
Comment (guozhen1997) [reply rlinf/models/embodiment/dreamzero/patch/wan_video_vae.py:32]: done

**Review by qurakchin — APPROVED** [review]

File: examples/sft/config/droid_sft_dreamzero.yaml (hunk 1 above)
Comment (thereAreDemonsNearby) [reply examples/sft/config/droid_sft_dreamzero.yaml:77]: @qurakchin Is this explaination reasonable enough?

**Review by thereAreDemonsNearby — APPROVED** [review]

**Review by andylin-hao — APPROVED** [review]
